### PR TITLE
[RFC] Fix NOEXCEPT_SCOPE (before it calls std::terminate and looses the exception)

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix NOEXCEPT_SCOPE (before it calls std::terminate and looses the exception)

Current implementation of NOEXCEPT_SCOPE will not work, you cannot
rethrow exception outside the catch block, this will simply terminate
(via std::terminate) the program.
In other words NOEXCEPT_SCOPE macro will simply call std::terminate on
exception and will lost original exception.

But if NOEXCEPT_SCOPE will accept the code that should be runned w/o
exceptions, then it can catch exception and log it, rewrite it in this
way.

Cc: @tavplubix

### CODE REVIEW

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 1]
```diff
@@ -4942,78 +4942,78 @@ MergeTreeData::DataPartsVector MergeTreeData::Transaction::commit(MergeTreeData:
             }
         }
 
-        NOEXCEPT_SCOPE;
+        NOEXCEPT_SCOPE({
```
Comment (tavplubix) [inline src/Storages/MergeTree/MergeTreeData.cpp:4945]: It's much more cumbersome. But maybe saved exception is worth it...

**Review by tavplubix — APPROVED** [review]

File: src/Storages/MergeTree/MergeTreeData.cpp (hunk 1 above)
Comment (azat) [reply src/Storages/MergeTree/MergeTreeData.cpp:4945]: >It's much more cumbersome

Indeed. But this makes the scope explicit, and for noexcept I think it is better to have explicit scope, and it should not be such huge like here. Also even if it should be such huge it can be moved into a separate function.
